cmd/admin-revoker/main.go

```diff
@@ -221,6 +372,108 @@ func (rc revocationCodes) Len() int           { return len(rc) }
+	// Attempt to find and decode the first PEM block.
+	block, _ := pem.Decode(keyContents)
+	if block == nil {
+		return nil, errors.New("does not contain a PEM formatted block")
+	}
+
+	// Attempt to parse the PEM block as a private key in a PKCS #1 container.
+	rsaKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err == nil {
+		return rsaKey, nil
+	}
+
+	// Attempt to parse the PEM block as a private key in a PKCS #8 container.
+	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err == nil {
+		switch k := key.(type) {
+		case *rsa.PrivateKey:
+			return k, nil
+
+		case *ecdsa.PrivateKey:
+			return k, nil
+		}
+	}
+
+	// Attempt to parse the PEM block as a private key in a SEC 1 container.
+	ecdsaKey, err := x509.ParseECPrivateKey(block.Bytes)
+	if err == nil {
+		return ecdsaKey, nil
+	}
+
+	// No private key could be parsed from the PEM block.
+	return nil, errors.New("cannot parse a private key from the provided PEM file")
+}
+
+// verifyRSAKeyPair is broken out of `verifyPrivateKey` for testing purposes.
+func verifyRSAKeyPair(privKey *rsa.PrivateKey, pubKey *rsa.PublicKey, msgHash hash.Hash) error {
+	signatureRSA, err := rsa.SignPSS(rand.Reader, privKey, crypto.SHA256, msgHash.Sum(nil), nil)
+	if err != nil {
+		return fmt.Errorf("failed to sign using the provided RSA private key: %s", err)
+	}
+
+	err = rsa.VerifyPSS(pubKey, crypto.SHA256, msgHash.Sum(nil), signatureRSA, nil)
+	if err != nil {
+		return fmt.Errorf("the provided ECDSA private key failed signature verification: %s", err)
+	}
+	return err
+}
+
+// verifyECDSAKeyPair is broken out of `verifyPrivateKey` for testing purposes.
+func verifyECDSAKeyPair(privKey *ecdsa.PrivateKey, pubKey *ecdsa.PublicKey, msgHash hash.Hash) error {
+	r, s, err := ecdsa.Sign(rand.Reader, privKey, msgHash.Sum(nil))
+	if err != nil {
+		return fmt.Errorf("failed to sign using the provided ECDSA private key: %s", err)
+	}
+
+	verify := ecdsa.Verify(pubKey, msgHash.Sum(nil), r, s)
+	if !verify {
+		return errors.New("the provided ECDSA private key failed signature verification")
+	}
+	return err
+}
+
+// verifyPrivateKey accepts a private key in the form of a `crypto.Signer` and
+// performs signing and verification for both RSA and ECDSA private keys.
+// Returns a nil error if signing and verification has been completed
+// successfully.
+func verifyPrivateKey(privateKey crypto.Signer) error {
+	msgHash := sha256.New()
+	_, err := msgHash.Write([]byte("verifiable"))
+	if err != nil {
+		return fmt.Errorf("failed to hash 'verifiable' message: %s", err)
+	}
+
+	privateKeyRSA, ok := privateKey.(*rsa.PrivateKey)
+	if ok {
+		return verifyRSAKeyPair(privateKeyRSA, &privateKeyRSA.PublicKey, msgHash)
+	} else {
+		privateKeyECDSA, ok := privateKey.(*ecdsa.PrivateKey)
+		if ok {
+			return verifyECDSAKeyPair(privateKeyECDSA, &privateKeyECDSA.PublicKey, msgHash)
+		}
+	}
+	// This should not happen.
+	return errors.New("the provided private key could not be asserted to ECDSA or RSA")
```

Comment:
A cool way to do this is:

```
			switch signer.(type) {
			case *rsa.PrivateKey:
				...
			case *ecdsa.PrivateKey:
				...
			default:
			        ...
```

Reply:
Ah, yeah a case switch much cleaner here.